### PR TITLE
exporter: add ceph_daemon labels to labeled counters as well

### PR DESCRIPTION
**Pull request by avanthakkar**
Fixes: https://tracker.ceph.com/issues/62874

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by idryomov — COMMENTED** [review]
The inclusion of "common: perf counters cache + rgw op labeled counters" commit in this PR is an accident, right?

**Comment by avanthakkar** [comment]
@idryomov @pereman2 I did test the PR with rgw metrcis which are getting introduced here https://github.com/ceph/ceph/pull/53003 and seems like it works fine, since all rgw metrics should have `instance_id` label similar to `ceph_daemon `  for other ceph daemons.

```
# HELP ceph_rgw_op_copy_obj_lat_count Copy object latency Count
# TYPE ceph_rgw_op_copy_obj_lat_count counter
ceph_rgw_op_copy_obj_lat_count{instance_id="mjdskk"} 0
ceph_rgw_op_copy_obj_lat_count{instance_id="shvvig"} 0
ceph_rgw_op_copy_obj_lat_count{Bucket="bkt",User="test3",instance_id="shvvig"} 0
ceph_rgw_op_copy_obj_lat_count{Bucket="bkt2",User="test3",instance_id="shvvig"} 0
ceph_rgw_op_copy_obj_lat_count{User="anonymous",instance_id="shvvig"} 0
ceph_rgw_op_copy_obj_lat_count{User="dashboard",instance_id="shvvig"} 0
```

File: src/exporter/DaemonMetricCollector.cc [hunk 1]
```diff
@@ -147,23 +147,24 @@ void DaemonMetricCollector::dump_asok_metrics() {
...
+
           // For now this is only required for rgw multi-site metrics
           auto multisite_labels_and_name = add_fixed_name_metrics(counter_name);
           if (!multisite_labels_and_name.first.empty()) {
             labels.insert(multisite_labels_and_name.first.begin(), multisite_labels_and_name.first.end());
             counter_name = multisite_labels_and_name.second;
           }
-          auto perf_values = counters_values.at(counter_name_init);
-          dump_asok_metric(counter_group, perf_values, counter_name, labels);
+          if (counters_values.find(counter_name_init) != counters_values.end()) {
```
Comment (idryomov) [inline src/exporter/DaemonMetricCollector.cc:164]: This change seems unrelated to the issue being addressed here -- what is this `if` condition for?

How can a counter which is present in `counters` (where `counter_name_init` comes from) be not present in `counters_values`?

**Comment by idryomov** [comment]
> @idryomov @pereman2 I did test the PR with rgw metrcis which are getting introduced here #53003 and seems like it works fine, since all rgw metrics should have `instance_id` label similar to `ceph_daemon ` for other ceph daemons.

Please paste before and after output for rbd-mirror daemon metrics (snapshot-based mirroring, at least two images with at least two mirror snapshots on each).

File: src/exporter/DaemonMetricCollector.cc (hunk 1 above)
Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc]: Yes it shouldn't include that change. Got those from previous cherry-picks. Removed them!

**Comment by avanthakkar** [comment]
> > @idryomov @pereman2 I did test the PR with rgw metrcis which are getting introduced here #53003 and seems like it works fine, since all rgw metrics should have `instance_id` label similar to `ceph_daemon ` for other ceph daemons.
> 
> Please paste before and after output for rbd-mirror daemon metrics (snapshot-based mirroring, at least two images with at least two mirror snapshots on each).

All new labeled metrics have `ceph_daemon` label now
https://paste.sh/UJUMR1ym#hCZATXI5EYsyPLkSV0DeUxhy

**Review by idryomov — COMMENTED** [review]
- The following in `get_labels_and_metric_name()` appears to be dead code -- rbd-mirror doesn't have any metrics that would start with `rbd_mirror_image_` anymore:
  ```
    if (daemon_name.find("rbd-mirror") != std::string::npos) {
      std::regex re(
          "^rbd_mirror_image_([^/]+)/(?:(?:([^/]+)/"
          ")?)(.*)\\.(replay(?:_bytes|_latency)?)$");
      std::smatch match;
      if (std::regex_search(daemon_name, match, re) == true) {
        new_metric_name = "ceph_rbd_mirror_image_" + match.str(4);
        labels["pool"] = quote(match.str(1));
        labels["namespace"] = quote(match.str(2));
        labels["image"] = quote(match.str(3));
      }
    }
  ```
- The commit message is a bit backwards:
  ```
    exporter: add ceph_daemon labels to labeled counters as well
    
    Fixes: https://tracker.ceph.com/issues/62874
    Signed-off-by: avanthakkar <avanjohn@gmail.com>
    
    Exporter missed adding the `ceph_daemon` or `instance_id`
    labels(in case if rgw metrics) to the new labeled performance counters.
  ```
  `Fixes` and `Signed-off-by` tags should be placed at the end, not in the middle.

**Comment by avanthakkar** [comment]
> * The following in `get_labels_and_metric_name()` appears to be dead code -- rbd-mirror doesn't have any metrics that would start with `rbd_mirror_image_` anymore:
>       ```
>         if (daemon_name.find("rbd-mirror") != std::string::npos) {
>           std::regex re(
>               "^rbd_mirror_image_([^/]+)/(?:(?:([^/]+)/"
>               ")?)(.*)\\.(replay(?:_bytes|_latency)?)$");
>           std::smatch match;
>           if (std::regex_search(daemon_name, match, re) == true) {
>             new_metric_name = "ceph_rbd_mirror_image_" + match.str(4);
>             labels["pool"] = quote(match.str(1));
>             labels["namespace"] = quote(match.str(2));
>             labels["image"] = quote(match.str(3));
>           }
>         }
>       ```
> 
>     * The commit message is a bit backwards:
>       ```
>         exporter: add ceph_daemon labels to labeled counters as well
>         
>         Fixes: https://tracker.ceph.com/issues/62874
>         Signed-off-by: avanthakkar <avanjohn@gmail.com>
>         
>         Exporter missed adding the `ceph_daemon` or `instance_id`
>         labels(in case if rgw metrics) to the new labeled performance counters.
>       ```
>       
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       `Fixes` and `Signed-off-by` tags should be placed at the end, not in the middle.

Ok! I 'd remove the rbd-mirror metrics regex.

File: src/exporter/DaemonMetricCollector.cc [hunk 2]
```diff
@@ -326,18 +325,6 @@ DaemonMetricCollector::get_labels_and_metric_name(std::string daemon_name,
   } else {
     labels.insert({"ceph_daemon", quote(daemon_name)});
   }
-  if (daemon_name.find("rbd-mirror") != std::string::npos) {
-    std::regex re(
-        "^rbd_mirror_image_([^/]+)/(?:(?:([^/]+)/"
-        ")?)(.*)\\.(replay(?:_bytes|_latency)?)$");
-    std::smatch match;
-    if (std::regex_search(daemon_name, match, re) == true) {
-      new_metric_name = "ceph_rbd_mirror_image_" + match.str(4);
```
Comment (idryomov) [inline src/exporter/DaemonMetricCollector.cc:335]: Nothing else sets `new_metric_name`, so `get_labels_and_metric_name()` can be changed to return only `labels_t` instead of a pair now.  I would also rename it to something like `get_extra_labels()` for clarity.

Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc:335]: This appears to be a cleanup and refactoring task related to the exporter. Would it be feasible to treat this as a distinct cleanup task? Also given we need to get this fix in ASAP

Comment (idryomov) [reply src/exporter/DaemonMetricCollector.cc:335]: This is a trivial refactor that should take a few minutes, so I don't see the point of delaying it.  Just make this PR consist of two commits: one for the original fix and one for the removal of the regex branch + change of signature + rename.

Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc:335]: Done

File: src/test/exporter/test_exporter.cc [hunk 3]
```diff
@@ -670,27 +670,25 @@ TEST(Exporter, check_labels_and_metric_name) {
   counters_data.emplace_back("ceph-osd.0", "ceph_osd_numpg");
   counters_data.emplace_back("ceph-client.rgw.foo.ceph-node-00.hrgsea.2.94739968030880", "ceph_rgw_get");
 
-  static std::vector<std::pair<labels_t, std::string>> labels_and_name;
-  labels_and_name.emplace_back(labels_t{{"ceph_daemon", "\"osd.0\""}}, "ceph_osd_numpg");
-  labels_and_name.emplace_back(labels_t{{"instance_id", "\"hrgsea\""}}, "ceph_rgw_get");
+  static std::vector<labels_t> labels_and_name;
```
Comment (idryomov) [inline src/test/exporter/test_exporter.cc:673]: There are no names here anymore:
```suggestion
  static std::vector<labels_t> labels_vec;
```

File: src/exporter/DaemonMetricCollector.cc [hunk 4]
```diff
@@ -147,15 +147,13 @@ void DaemonMetricCollector::dump_asok_metrics() {
...
 
-          if (counters_labels.empty()) {
-            auto labels_and_name = get_labels_and_metric_name(daemon_name, counter_name);
-            if (labels_and_name.first.empty()) {
-              dout(1) << "Unable to parse instance_id from daemon_name: " << daemon_name << dendl;
-              continue;
-            }
-            labels = labels_and_name.first;
-            counter_name = labels_and_name.second;
+          auto labels_ = get_extra_labels(daemon_name);
```
Comment (idryomov) [inline src/exporter/DaemonMetricCollector.cc:150]: An automatic variable ending with an underscore is weird -- a trailing underscore designates class members in some style guides:
```suggestion
          auto extra_labels = get_extra_labels(daemon_name);
```

File: src/test/exporter/test_exporter.cc [hunk 5]
```diff
@@ -670,27 +670,25 @@ TEST(Exporter, check_labels_and_metric_name) {
...
-  static std::vector<std::pair<labels_t, std::string>> labels_and_name;
-  labels_and_name.emplace_back(labels_t{{"ceph_daemon", "\"osd.0\""}}, "ceph_osd_numpg");
-  labels_and_name.emplace_back(labels_t{{"instance_id", "\"hrgsea\""}}, "ceph_rgw_get");
+  static std::vector<labels_t> labels_vec;
+  labels_vec.emplace_back(labels_t{{"ceph_daemon", "\"osd.0\""}});
+  labels_vec.emplace_back(labels_t{{"instance_id", "\"hrgsea\""}});
   auto counter_data_itr = counters_data.begin();
-  auto labels_and_name_itr = labels_and_name.begin();
-  for (; counter_data_itr != counters_data.end() && labels_and_name_itr != labels_and_name.end();
+  auto labels_and_name_itr = labels_vec.begin();
```
Comment (idryomov) [inline src/test/exporter/test_exporter.cc:677]: Again, there are no names here anymore:
```suggestion
  auto labels_vec_itr = labels_vec.begin();
```

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
jenkins test make check

**Comment by idryomov** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
Dashboard e2e and cephadm e2e failures aren't related to the PR. So it's ready to merge. @idryomov  You mind taking care of merging the PR?
